contracts/ERC1155NftCustom.sol

```diff
@@ -0,0 +1,188 @@
+    address private _owner;
+
+    bool public metadataUpdatable;
+    bool public tokensBurnable;
+    bool public tokensTransferable;
+
+    string public name;
+    string public symbol;
+    string public baseURI;
+
+    mapping (uint256 => bool) public freezeTokenUris;
+    mapping (uint256 => uint256) public tokenSupply;
+    mapping(uint256 => string) private _tokenURIs;
+
+    event PermanentURI(string _value, uint256 indexed _id); // https://docs.opensea.io/docs/metadata-standards
+    event PermanentURIGlobal();
+
+    constructor(
+        string memory _name, 
+        string memory _symbol, 
+        address owner, 
+        bool _metadataUpdatable, 
+        bool _tokensBurnable,
+        bool _tokensTransferable,
+        string memory _initBaseURI,
+        string memory _defaultUri
+    ) ERC1155(_defaultUri) {
+        _setupRole(DEFAULT_ADMIN_ROLE, owner);
+        _setupRole(MINTER_ROLE, owner);
+        _setupRole(MINTER_ROLE, msg.sender);
+
+        metadataUpdatable = _metadataUpdatable;
+        tokensBurnable = _tokensBurnable;
+        tokensTransferable = _tokensTransferable;
+
+        baseURI = _initBaseURI;
+        _owner = owner;
+        name = _name;
+        symbol = _symbol;
+    }
+
+    function setURI(string memory _newURI) public onlyRole(MINTER_ROLE) {
+        _setURI(_newURI);
+    }
+
+    function updateTokenUri(uint256 _tokenId, string memory _newUri, bool _isFreezeTokenUri)
+    public
+    onlyRole(MINTER_ROLE) {
+        require(_exists(_tokenId), "NFT: update URI query for nonexistent token");
+        require(metadataUpdatable, "NFT: Token uris are frozen globally");
+        require(freezeTokenUris[_tokenId] != true, "NFT: Token is frozen");
+        require(_isFreezeTokenUri || (bytes(_newUri).length != 0), "NFT: Either _newUri or _isFreezeTokenUri=true required");
+
+        if (bytes(_newUri).length != 0) {
+            require(keccak256(bytes(_tokenURIs[_tokenId])) != keccak256(bytes(string(abi.encodePacked(_newUri)))), "NFT: New token URI is same as updated");
+            _tokenURIs[_tokenId] = _newUri;
+            emit URI(_newUri, _tokenId);
+        }
+        if (_isFreezeTokenUri) {
+            freezeTokenUris[_tokenId] = true;
+            emit PermanentURI(_tokenURIs[_tokenId], _tokenId);
+        }
+    }
+
+    function burn(
+        uint256 id,
+        uint256 value
+    ) public onlyRole(MINTER_ROLE) {
+        require(tokensBurnable, "NFT: tokens burning is disabled");
+        
+        _burn(_owner, id, value);
+        tokenSupply[id] -= value;
+    }
+
+    function burnBatch(
+        uint256[] memory ids,
+        uint256[] memory values
+    ) public onlyRole(MINTER_ROLE) {
+        require(tokensBurnable, "NFT: tokens burning is disabled");
+        _burnBatch(_owner, ids, values);
+        for (uint256 i = 0; i < ids.length; i++) {
+            tokenSupply[ids[i]] -= values[i];
+        }
+    }
+
+    function transferByOwner(
```

Comment:
Is this meant to transfer tokens only from the contract owner to another address or between any two addresses?

Reply:
only the _owner